### PR TITLE
feat: tfidf encoder

### PR DESCRIPTION
**Pull request by davidbp**
Add a simple TF-IDF encoder as an example that displays sparse support in Jina.

### CODE REVIEW

File: encoders/nlp/TFIDFTextEncoder/__init__.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    """
+
+    def __init__(self,
+                 path_vectorizer= "./pods/tfidf_vectorizer.pickle",
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.path_vectorizer = path_vectorizer
+
+    def post_init(self):
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/__init__.py:24]: call super `post_init`

File: encoders/nlp/TFIDFTextEncoder/__init__.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
+__copyright__ = "Copyright (c) 2021 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+import numpy as np
+import torch
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/__init__.py:5]: remember to not add `external imports` on top

File: encoders/nlp/TFIDFTextEncoder/__init__.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                 path_vectorizer= "./pods/tfidf_vectorizer.pickle",
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.path_vectorizer = path_vectorizer
+
+    def post_init(self):
+        self.tfidf_vectorizer = pickle.load(open(self.path_vectorizer, "rb"))
+
+    def encode(self, data, *args, **kwargs) -> 'np.ndarray':
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/__init__.py:27]: what is the real signature right now?

Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/__init__.py]: updated

File: encoders/nlp/TFIDFTextEncoder/__init__.py [hunk 4]
```diff
@@ -15,14 +16,15 @@ class TFIDFTextEncoder(BaseEncoder):
...
+                 path_vectorizer= "./model/tfidf_vectorizer.pickle",
                  *args,
                  **kwargs):
         super().__init__(*args, **kwargs)
         self.path_vectorizer = path_vectorizer
 
     def post_init(self):
-        self.tfidf_vectorizer = pickle.load(open(self.path_vectorizer, "rb"))
+        import pickle
+        self.tfidf_vectorizer = pickle.load(open( os.path.join(cur_dir,self.path_vectorizer), "rb"))
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/__init__.py:27]: just assume the `path_vectorizer` is absolute, why should someone use the `code` folder as a base for storing files? Also use `single quote`

File: encoders/nlp/TFIDFTextEncoder/__init__.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.path_vectorizer = path_vectorizer
+
+    def post_init(self):
+        import pickle
+        self.tfidf_vectorizer = pickle.load(open( os.path.join(cur_dir,self.path_vectorizer), "rb"))
+
+    def encode(self, data, *args, **kwargs) -> 'scipy.sparse.csr_matrix':
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/__init__.py:29]: can we use a `custom` batching decorator (independent from the one in the core) that ensures that `data` arrives in `batches` and that the results are merged from a single `csr_matrix`?

File: encoders/nlp/TFIDFTextEncoder/__init__.py (hunk 4 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/__init__.py]: true, done

File: encoders/nlp/TFIDFTextEncoder/__init__.py (hunk 5 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/__init__.py:26]: We could but all decorators from core are for numpy arrays, not sparse arrays (we should construct it then, but core does not depend on scipy making this a bit infeasible in core)

File: encoders/nlp/TFIDFTextEncoder/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,16 @@
+FROM jinaai/jina:1.1.0
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/Dockerfile:1]: remove version here as this update is handled by dependabot

File: encoders/nlp/TFIDFTextEncoder/__init__.py (hunk 1 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/__init__.py:22]: Don't know what you mean...

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoders/nlp/TFIDFTextEncoder/__init__.py (hunk 1 above)
Comment (JoanFM) [reply encoders/nlp/TFIDFTextEncoder/__init__.py:22]: u need to call the post init of the super class.

File: encoders/nlp/TFIDFTextEncoder/__init__.py (hunk 5 above)
Comment (JoanFM) [reply encoders/nlp/TFIDFTextEncoder/__init__.py:26]: thats the point, have it here and custom for csr matrix

File: encoders/nlp/TFIDFTextEncoder/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
+# TFIDFTextEncoder
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/README.md:1]: Changes in the `README.md`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoders/nlp/TFIDFTextEncoder/README.md (hunk 7 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/README.md:1]: updated

File: encoders/nlp/TFIDFTextEncoder/README.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+def load_data():
+    from sklearn.datasets import fetch_20newsgroups
+    newsgroups_train = fetch_20newsgroups(subset='train')
+    return newsgroups_train.data
+
+if __name__ == '__main__':
+    X = load_data()    
+    tfidf_vectorizer = TfidfVectorizer()
+    tfidf_vectorizer.fit(X)
+    pickle.dump(tfidf_vectorizer, open("tfidf_vectorizer.pickle", "wb"))
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/README.md:47]: I need information about how to use the `TFIDFTextEncoder` and not the `TFIDFVectorizer`.

You can show how from this pickle u can load a `TFIDFTextEncoder`

Comment (JoanFM) [reply encoders/nlp/TFIDFTextEncoder/README.md:47]: please use single quote

File: encoders/nlp/TFIDFTextEncoder/__init__.py (hunk 1 above)
Comment (JoanFM) [reply encoders/nlp/TFIDFTextEncoder/__init__.py:22]:
```suggestion
    def post_init(self):
         super().post_init()
```

File: encoders/nlp/TFIDFTextEncoder/__init__.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+                 *args,
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self.path_vectorizer = path_vectorizer
+
+    def post_init(self):
+        import pickle
+        self.tfidf_vectorizer = pickle.load(open(self.path_vectorizer, "rb"))
+
+    def encode(self, data, *args, **kwargs) -> 'scipy.sparse.csr_matrix':
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/__init__.py:26]: please add type annotations for `data`

File: encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    
+    # fit text featurizer descriptor
+    tfidf_vectorizer = TfidfVectorizer()
+    tfidf_vectorizer.fit(X)
+
+    # store the object to disk
+    pickle.dump(tfidf_vectorizer, open("tfidf_vectorizer.pickle", "wb"))
+    
+    # load the object later
+    # tfidf_vectorizer = pickle.load(open("tfidf_vectorizer.pickle", "rb"))
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py:23]: remove line

File: encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
+from sklearn.feature_extraction.text import TfidfVectorizer
+import csv
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py:2]: this script is not needed?

File: encoders/nlp/TFIDFTextEncoder/tests/encoding_with_original_tfidf.py [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import os
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/tests/encoding_with_original_tfidf.py:1]: one of this is not needed I guess

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py (hunk 11 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py:2]: It's needed to genereta the dummy pickle that we can include for tests, just as a helper

Comment (JoanFM) [reply encoders/nlp/TFIDFTextEncoder/model/build_tfidf.py:2]: one of the 2 seems innecessary

File: encoders/nlp/TFIDFTextEncoder/tests/encoding_with_original_tfidf.py (hunk 12 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/tests/encoding_with_original_tfidf.py:1]: right

File: encoders/nlp/TFIDFTextEncoder/tests/test_tfidftextencoder.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+    # Encoder embedding 
+    encoder = TFIDFTextEncoder()
+
+    print_array_info(text_batch, 'text_batch')
+    embeddeding_batch = encoder.encode(text_batch)
+    print_array_info(embeddeding_batch, 'embeddeding_batch')
+
+    # Compare with ouptut 
+    expected_batch = scipy.sparse.load_npz(os.path.join(cur_dir, 'expected_batch.npz'))
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/tests/test_tfidftextencoder.py:45]: can u ensure the `shape`? meaning that it has as many rows as input `sentences` in the `batch`?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoders/nlp/TFIDFTextEncoder/tests/test_tfidftextencoder.py (hunk 13 above)
Comment (davidbp) [reply encoders/nlp/TFIDFTextEncoder/tests/test_tfidftextencoder.py:45]: I think it is implicit in the matrix detection but I will make it explicit

File: encoders/nlp/TFIDFTextEncoder/README.md [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+- Input shape: `BatchSize `
+
+- Output shape: `BatchSize x EmbeddingDimension`
+
+
+
+
+## How to 
+
+### Dependencies
```
Comment (JoanFM) [inline encoders/nlp/TFIDFTextEncoder/README.md:18]: Is this necessary @rutujasurve94 ?

**Review by JoanFM — APPROVED** [review]

**Pull request merged**